driver-core/src/main/com/mongodb/internal/connection/ScramShaAuthenticator.java

```diff
@@ -82,9 +85,33 @@ protected void appendSaslStartOptions(final BsonDocument saslStartCommand) {
 
     @Override
     protected SaslClient createSaslClient(final ServerAddress serverAddress) {
+        if (saslClient != null && this.speculativeAuthenticateResponse != null) {
+            return saslClient;
+        }
         return new ScramShaSaslClient(getMongoCredentialWithCache(), randomStringGenerator, authenticationHashGenerator);
     }
 
+    @Override
+    protected BsonDocument createSpeculativeAuthenticateCommand(final InternalConnection connection) {
+        try {
+            saslClient = createSaslClient(connection.getDescription().getServerAddress());
+            return createSaslStartCommandDocument(saslClient.evaluateChallenge(new byte[0]))
```

Comment:
Also need to call `appendSaslStartOptions` here.  Not sure if this is enough of an issue to require a refactoring to make it harder to forget to do this.  Think about it.